lib/browser/api/web-contents.js

```diff
@@ -160,6 +160,20 @@ WebContents.prototype.executeJavaScript = function (code, hasUserGesture, callba
   }
 }
 
+WebContents.prototype.takeHeapSnapshot = function (...args) {
+  return new Promise((resolve, reject) => {
+    const requestId = getNextId()
+    this.send('ELECTRON_RENDERER_TAKE_HEAP_SNAPSHOT', requestId, args)
+    ipcMain.once(`ELECTRON_BROWSER_TAKE_HEAP_SNAPSHOT_RESULT_${requestId}`, (event, error, result) => {
```

Comment:
I know the chances of this happening in this particular case are basically 0 but to avoid race conditions it's usually better to add the listener before making the request.  E.g.

```js
ipcMain.on('response', doThing)
sendRequest()
```

Reply:
I'll do it here. However there are lots of places in the code where we don't do it properly.

Reply:
done

Reply:
> However there are lots of places in the code where we don't do it properly.

Yeah I don't think the catch us out because IPC is basically guaranteed to not be in the same tick, but we should address those to handle that case 🤔